### PR TITLE
impl(otel): trace context across AsyncPollingLoop

### PR DESCRIPTION
**Pull request by dbolduc**
Part of the work for #10618 

Propagate trace context across an `Async(Rest)PollingLoop`

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11040)

### CODE REVIEW

File: google/cloud/internal/async_polling_loop.cc [hunk 1]
```diff
@@ -44,13 +45,14 @@ class AsyncPollingLoopImpl
...
-      if (auto self = w.lock()) {
-        OptionsSpan span(std::move(options));
-        self->DoCancel();
-      }
-    });
+    CallContext call_context;
+    promise_ = promise<StatusOr<Operation>>(
+        [w, c = std::move(call_context)]() mutable {
+          if (auto self = w.lock()) {
+            ScopedCallContext scope(c);
```
Comment (devbww) [inline google/cloud/internal/async_polling_loop.cc:52]: Move the `CallContext` into the `ScopedCallContext`?

File: google/cloud/internal/async_rest_polling_loop.cc [hunk 2]
```diff
@@ -45,13 +46,14 @@ class AsyncRestPollingLoopImpl
...
-    auto const& options = internal::CurrentOptions();
-    promise_ = promise<StatusOr<Operation>>([w, options]() mutable {
-      if (auto self = w.lock()) {
-        internal::OptionsSpan span(std::move(options));
-        self->DoCancel();
-      }
-    });
+    internal::CallContext call_context;
+    promise_ = promise<StatusOr<Operation>>(
+        [w, c = std::move(call_context)]() mutable {
```
Comment (coryan) [inline google/cloud/internal/async_rest_polling_loop.cc:51]: As above...

File: google/cloud/internal/async_polling_loop_test.cc [hunk 3]
```diff
@@ -618,6 +622,74 @@ TEST(AsyncPollingLoopTest, ConfigurePollContext) {
...
+  auto mock_cq = std::make_shared<MockCompletionQueueImpl>();
+  EXPECT_CALL(*mock_cq, MakeRelativeTimer)
+      .Times(AtLeast(1))
+      .WillRepeatedly([&](std::chrono::nanoseconds) {
+        EXPECT_THAT(span, IsActive());
+        return timer_sequencer.PushBack();
+      });
+  CompletionQueue cq(mock_cq);
+
+  AsyncSequencer<StatusOr<Operation>> get_sequencer;
```
Comment (devbww) [inline google/cloud/internal/async_polling_loop_test.cc:646]: This file looks like it is schizophrenic regarding the `google::longrunning::Operation`  using declaration.

File: google/cloud/internal/async_rest_polling_loop_test.cc [hunk 4]
```diff
@@ -561,6 +565,75 @@ TEST(AsyncRestPollingLoopTest, PollThenCancelDuringPoll) {
...
+  auto mock_cq = std::make_shared<MockCompletionQueueImpl>();
+  EXPECT_CALL(*mock_cq, MakeRelativeTimer)
+      .Times(AtLeast(1))
+      .WillRepeatedly([&](std::chrono::nanoseconds) {
+        EXPECT_THAT(span, IsActive());
+        return timer_sequencer.PushBack();
+      });
+  CompletionQueue cq(mock_cq);
+
+  AsyncSequencer<StatusOr<Operation>> get_sequencer;
```
Comment (devbww) [inline google/cloud/internal/async_rest_polling_loop_test.cc:589]: Ditto.

File: google/cloud/internal/async_polling_loop.cc [hunk 5]
```diff
@@ -44,13 +45,14 @@ class AsyncPollingLoopImpl
...
-    auto const& options = CurrentOptions();
-    promise_ = promise<StatusOr<Operation>>([w, options]() mutable {
-      if (auto self = w.lock()) {
-        OptionsSpan span(std::move(options));
-        self->DoCancel();
-      }
-    });
+    CallContext call_context;
+    promise_ = promise<StatusOr<Operation>>(
+        [w, c = std::move(call_context)]() mutable {
```
[the remark this answers is not in the document]
Comment (dbolduc) [reply google/cloud/internal/async_polling_loop.cc]: Fixed.

File: google/cloud/internal/async_rest_polling_loop.cc (hunk 2 above)
Comment (dbolduc) [reply google/cloud/internal/async_rest_polling_loop.cc]: As above.

File: google/cloud/internal/async_polling_loop.cc (hunk 1 above)
Comment (dbolduc) [reply google/cloud/internal/async_polling_loop.cc]: Fixed.

File: google/cloud/internal/async_polling_loop_test.cc (hunk 3 above)
Comment (dbolduc) [reply google/cloud/internal/async_polling_loop_test.cc]: Let's just spell the thing out. (I am not doing this in `async_(rest_)polling_loop.cc`, though.)

File: google/cloud/internal/async_rest_polling_loop_test.cc (hunk 4 above)
Comment (dbolduc) [reply google/cloud/internal/async_rest_polling_loop_test.cc]: Ditto.

**Review by devbww — APPROVED** [review]

**Pull request merged**